### PR TITLE
Using a common dialog box for the app

### LINKED ISSUE #527: Use a common dialog box for the app

### PR DESCRIPTION
**Pull request by dalmia**
Fixes #527 

## Summary
- removes the dialog box components from Editor and PlioListItem
- adds a dialog box component to App.vue to be used globally by each component
- creates a new store module for managing all the dialog states
- uses two new variables to mark whether the confirm/cancel button of the dialog has been clicked
- the actions to perform upon confirming/cancelling are defined within the component invoking the dialog box
- changed the margin top for the language selector to make it vertically centered
- used the store variable `selectedPlioId` instead of `plioId` for deleting plio - the reason has been mentioned as a comment
- defined helper functions within `tests` and reused them throughout the code (hence, removed `services/Testing/Utilities.js`)
- used nested `describe` within tests to reuse code across tests of the same file
- changed some `var`s to `let`s 

## Test Plan

- [x] Test Responsiveness
   - [ ] Laptop (1200px)
   - [ ] Tablet (760px)
   - [ ] Phone (320px)
- [x] Cross-Browser Testing
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari (use BrowserStack)
- [ ] ~Local Language Support~
- [x] Wrote tests
- [x] Tested locally
- [x] Self-review
- [x] Comments have been added appropriately
- [ ] ~Check for bundle size [here](https://bundlephobia.com/) if adding a package~
- [ ] Added relevant details like Labels/Projects/Milestones etc.
- [ ] ~Images have `alt` attributes~
- [ ] ~Any `` tags have `width` and `height` specified~
- [ ] ~Any `target="_blank"` links have `rel="noopener"`~
- [ ] ~Only SVGs are used as images. If PNGs are used, their size has been optimised.~
- [ ] ~Any SVGs without text have their `aria-label` attributes set~
- [ ] Tested on staging
- [ ] Tested on an actual physical phone
- [ ] Tested on production

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -437,7 +494,6 @@ export default {
           }
         })
         .catch(() => this.toast.error(this.$t("error.create_plio")));
-      this.resetMenuState();
```
Comment (dalmia) [inline src/App.vue:440]: removed this to ensure smoother transition from home page to editor

**Comment by dalmia** [comment]
Seems like deepscan is going to give a pain whenever we use helper functions in tests. We can disable deepscan on the tests folder too!

File: tests/unit/App.spec.js [hunk 2]
```diff
@@ -44,8 +59,91 @@ describe("App.vue", () => {
...
     // After this line, router is ready
     await router.isReady();
+  });
 
+  it("should render", async () => {
     expect(wrapper.vm.isAuthenticated).toBeTruthy();
-    console.log(wrapper.vm.pending);
+  });
+
+  it("shows dialog box when dialog box is set to be shown", async () => {
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply tests/unit/App.spec.js:68]: or properly.
`when dialog box is set to be shown` seemed redundant

File: tests/unit/App.spec.js [hunk 3]
```diff
@@ -44,8 +59,91 @@ describe("App.vue", () => {
...
+    // dialog box shouldn't be shown at first
+    expect(wrapper.vm.isDialogBoxShown).toBeFalsy();
+    expect(wrapper.find('[data-test="dialogBox"]').exists()).toBeFalsy();
+
+    // set the dialog box to be shown and its properties
```
Comment (deepansh96) [inline tests/unit/App.spec.js:73]:
```suggestion
    // set the properties of the dialog box and show it
```

File: tests/unit/App.spec.js [hunk 4]
```diff
@@ -44,8 +59,91 @@ describe("App.vue", () => {
...
+  describe("dialog box", () => {
+    beforeEach(async () => {
+      await store.dispatch("dialog/setDialogTitle", dialogTitle);
+      await store.dispatch("dialog/setDialogDescription", dialogDescription);
+      await store.dispatch(
+        "dialog/setConfirmButtonConfig",
+        confirmButtonConfig
+      );
+      await store.dispatch("dialog/setCancelButtonConfig", cancelButtonConfig);
+      await store.dispatch("dialog/setDialogBox");
```
Comment (deepansh96) [inline tests/unit/App.spec.js:106]: Why are these `await`? 
None of the actions are actually returning a promise, so no need to await for them

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 5]
```diff
@@ -1,28 +1,30 @@
...
 import PlioListItem from "@/components/Collections/ListItems/PlioListItem";
 import store from "@/store";
-import { setMatchMedia } from "@/services/Testing/Utilities";
 import mockAxios from "jest-mock-axios";
 
 describe("PlioListItem.vue", () => {
+  let wrapper;
+
   beforeEach(async () => {
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:10]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 6]
```diff
@@ -1,28 +1,30 @@
...
   it("should render with default values", () => {
-    const wrapper = mount(PlioListItem);
+    wrapper = mount(PlioListItem);
     expect(wrapper).toBeTruthy();
   });
 
   it("should load plio given plio id", () => {
     const loadPlio = jest.spyOn(PlioListItem.methods, "loadPlio");
-    mount(PlioListItem, {
+    wrapper = mount(PlioListItem, {
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:27]: no need of wrapper here no?

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 7]
```diff
@@ -173,7 +175,7 @@ describe("PlioListItem.vue", () => {
         },
       },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:179]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 8]
```diff
@@ -219,7 +221,7 @@ describe("PlioListItem.vue", () => {
         };
       },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:225]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 9]
```diff
@@ -264,7 +266,7 @@ describe("PlioListItem.vue", () => {
         },
       },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:270]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 10]
```diff
@@ -336,7 +338,7 @@ describe("PlioListItem.vue", () => {
         };
       },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:342]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 11]
```diff
@@ -373,7 +375,7 @@ describe("PlioListItem.vue", () => {
...
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:379]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 12]
```diff
@@ -474,15 +476,15 @@ describe("PlioListItem.vue", () => {
 
   it("clicking analyze routes to Dashboard ", async () => {
     // set `matches` as `True` for testing on touch screen devices
-    setMatchMedia(true);
+    setMatchMedia(true); // deepscan-disable-line
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:479]: Does setting `setMatchMedia` as true means that if any of our tests later on will require testing on a non-touch screen device, we'll have to set it back again to false?
Wouldn't it be better if we create a separate describe for touch devices and a separate describe for non-touch devices?

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 13]
```diff
@@ -500,7 +502,7 @@ describe("PlioListItem.vue", () => {
         },
       },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:506]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 14]
```diff
@@ -543,17 +540,12 @@ describe("PlioListItem.vue", () => {
...
       },
-      global: {
-        mocks: {
-          $router: mockRouter,
-        },
-      },
     });
-    // passing in plioID triggers startLoading which keeps the component in pending state
+    // passing in plioId triggers startLoading which keeps the component in pending state
     await store.dispatch("sync/stopLoading");
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:545]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 15]
```diff
@@ -590,13 +577,8 @@ describe("PlioListItem.vue", () => {
...
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:582]: no need of await here

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 16]
```diff
@@ -764,59 +723,105 @@ describe("PlioListItem.vue", () => {
...
+    // dialog confirm button state should have been reset
+    expect(wrapper.vm.isDialogConfirmClicked).toBeFalsy();
+    // dialog action should be unset
+    expect(wrapper.vm.dialogAction).toBeFalsy();
   });
 
-  it("delete confirmation dialog box margin is set correctly ", async () => {
-    // margin value changes based on window width
-    const wrapper = mount(PlioListItem, {
+  it("does not interfere with irrelevant dialog confirm trigger", async () => {
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:732]: description not clear!

File: tests/unit/components/Collections/ListItems/PlioListItem.spec.js [hunk 17]
```diff
@@ -764,59 +723,105 @@ describe("PlioListItem.vue", () => {
...
-    store.dispatch("generic/setWindowInnerWidth", 350);
-    await flushPromises();
-    expect(wrapper.get('[data-test="dialogBox"]').attributes("style")).toEqual(
-      "left: 15%;"
-    );
+    // the dialog action shouldn't have been affected and
+    // the confirm click status should remain active
+    expect(wrapper.vm.isDialogConfirmClicked).toBeTruthy();
+    expect(wrapper.vm.dialogAction).toBe(newDialogAction);
+  });
```
Comment (deepansh96) [inline tests/unit/components/Collections/ListItems/PlioListItem.spec.js:778]: Didn't understand

File: tests/unit/pages/Editor.spec.js [hunk 18]
```diff
@@ -73,30 +65,16 @@ describe("Editor.vue", () => {
...
-    expect(wrapper.vm.dialogTitle).toBe("Publishing the plio...");
-    expect(wrapper.vm.publishInProgressDialogTitle).toBe(
-      "Publishing the changes.."
-    );
-    expect(wrapper.vm.dialogConfirmButtonConfig.enabled).toBeFalsy();
-    expect(wrapper.vm.dialogConfirmButtonConfig.text).toBe("");
-    expect(wrapper.vm.dialogConfirmButtonConfig.class).toBe("");
-    expect(wrapper.vm.dialogCancelButtonConfig.enabled).toBeFalsy();
-    expect(wrapper.vm.dialogCancelButtonConfig.text).toBe("");
-    expect(wrapper.vm.dialogCancelButtonConfig.class).toBe("");
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:94]: Not checking these?

File: tests/unit/pages/Editor.spec.js [hunk 19]
```diff
@@ -1476,6 +1418,9 @@ describe("Editor.vue", () => {
     const itemEditorWrapper = wrapper.findComponent(ItemEditor);
     const inputTextWrapper = itemEditorWrapper.findAllComponents(InputText)[4];
 
+    // clear past values of dialog description
+    await store.dispatch("dialog/unsetDialogDescription");
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1422]: no need of await here
- Also, check and change in the codebase wherever non async store action dispatches are written using await

File: tests/unit/pages/Editor.spec.js [hunk 20]
```diff
@@ -1485,51 +1430,109 @@ describe("Editor.vue", () => {
...
+
+    // simulate clicking the confirm button of the dialog box
+    await simulateConfirmClick(); // deepscan-disable-line
+
+    // dialogAction and dialog confirm click status should be reset
+    expect(wrapper.vm.dialogAction).toBeFalsy();
+    expect(wrapper.vm.isDialogConfirmClicked).toBeFalsy();
+  });
+
+  it("cancelling delete option dialog resets index to delete", async () => {
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1484]:
```suggestion
  it("cancelling delete option dialog resets option's index to delete", async () => {
```
Was wondering for a while what index

File: tests/utils/dialog.js [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+global.simulateConfirmClick = async () => {
+  // simulate clicking the confirm button of the dialog box
+  await store.dispatch("dialog/unsetDialogBox");
+  await store.dispatch("dialog/unsetDialogTitle");
+  await store.dispatch("dialog/unsetDialogDescription");
+  await store.dispatch("dialog/unsetConfirmButtonConfig");
+  await store.dispatch("dialog/unsetCancelButtonConfig");
+  await store.dispatch("dialog/unsetDialogBoxClass");
+  await store.dispatch("dialog/unsetDialogCloseButton");
+  await store.dispatch("dialog/setConfirmClicked");
```
Comment (deepansh96) [inline tests/utils/dialog.js:12]: I guess no need of await here as well

File: tests/utils/dialog.js [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+global.simulateCancelClick = async () => {
+  // simulate clicking the cancel button of the dialog box
+  await store.dispatch("dialog/unsetDialogBox");
+  await store.dispatch("dialog/unsetDialogTitle");
+  await store.dispatch("dialog/unsetDialogDescription");
+  await store.dispatch("dialog/unsetConfirmButtonConfig");
+  await store.dispatch("dialog/unsetCancelButtonConfig");
+  await store.dispatch("dialog/unsetDialogBoxClass");
+  await store.dispatch("dialog/unsetDialogCloseButton");
+  await store.dispatch("dialog/setCancelClicked");
```
Comment (deepansh96) [inline tests/utils/dialog.js:24]: I guess no need of await here as well

File: tests/unit/pages/Editor.spec.js [hunk 23]
```diff
@@ -1841,4 +1843,60 @@ describe("Editor.vue", () => {
...
+
+    await wrapper.find('[data-test="publishButton"]').trigger("click");
+
+    // change the dialog action so that it is no longer
+    // relevant to this component
+    const newDialogAction = "testAction";
+    await store.dispatch("dialog/setDialogAction", newDialogAction);
+
+    // simulate clicking the cancel button of the dialog box
+    await simulateCancelClick(); // deepscan-disable-line
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1892]: I think you won't need to write this `// deepscan-disable-line` if you reference values and objects using `global`.
Like
`await global.simulateCancelClick()`

I had the same issue while implementing `Offline.js`
So importing the `Offline` npm package adds it to the global `window` object. All the methods are accessible by either using `window.checkOnlineStatus()` or directly like this `checkOnlineStatus()`.

Using the latter one gave me the same deepscan error, so I used `window.checkOnlineStatus()`. And this solution makes more sense as it's obvious to the reader that this is a method from the global scope so don't keep looking for this method in the current file

File: tests/utils/dummyData.js [hunk 24]
```diff
@@ -427,33 +427,33 @@ export const imageData = {
...
+global.dummyAccessToken = {
   access_token: "1234",
   expires_in: 86400,
   refresh_token: "4321",
   scope: "read write",
   token_type: "Bearer",
 };
 
-export let dummyItemResponses = [
+global.dummyItemResponses = [
```
Comment (deepansh96) [inline tests/utils/dummyData.js:456]: Just curious, how does the compiler know what type of variable this is? How does it know if it's a var, a let, or a const?

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: tests/utils/dummyData.js (hunk 24 above)
Comment (dalmia) [reply tests/utils/dummyData.js:456]: given it is being added to the global object, it is likely global scope (hence, var) and not block scope!

File: src/App.vue [hunk 25]
```diff
@@ -500,10 +558,20 @@ export default {
...
     ...mapState("sync", ["pending"]),
+    ...mapState("dialog", {
+      dialogTitle: (state) => state.title,
+      dialogDescription: (state) => state.description,
+      isDialogBoxShown: (state) => state.isShown,
+      isDialogCloseButtonShown: (state) => state.isCloseButtonShown,
+      dialogConfirmButtonConfig: (state) => state.confirmButtonConfig,
+      dialogCancelButtonConfig: (state) => state.cancelButtonConfig,
+      dialogBoxClass: (state) => state.boxClass,
+      dialogAction: (state) => state.action,
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply src/App.vue:574]: Change at all places

File: src/components/Collections/ListItems/PlioListItem.vue [hunk 26]
```diff
@@ -261,13 +234,61 @@ export default {
...
+  watch: {
+    isDialogConfirmClicked(value) {
+      if (value) {
+        switch (this.dialogAction) {
+          case "deletePlio":
+            this.deletePlio();
+            break;
+          default:
+            // returning here so that it doesn't interfere with the
+            // confirmation step of a different dialog action
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Collections/ListItems/PlioListItem.vue:246]: different dialogAction

File: src/components/Collections/ListItems/PlioListItem.vue [hunk 27]
```diff
@@ -261,13 +234,61 @@ export default {
...
+            return;
+        }
+        this.unsetConfirmClicked();
+        // reset the dialog action value if no new dialog
+        // boxes have been shown
+        if (!this.isDialogBoxShown) this.unsetDialogAction();
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Collections/ListItems/PlioListItem.vue:252]: write negation:
- if some new dialog box is created from one of the actions above, we don't want to reset the dialog action

File: src/pages/Editor.vue [hunk 28]
```diff
@@ -973,7 +1016,6 @@ export default {
     isBackgroundDisabled() {
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Editor.vue:1016]: will be resolved after merging this: https://github.com/avantifellows/plio-frontend/pull/536

File: tests/unit/pages/Editor.spec.js (hunk 23 above)
Comment (dalmia) [reply tests/unit/pages/Editor.spec.js:1892]: do this for dummyItems, etc.

File: src/pages/Editor.vue [hunk 29]
```diff
@@ -708,6 +683,68 @@ export default {
...
+        switch (this.dialogAction) {
+          case "deleteOption":
+            this.cancelDeleteOption();
+            break;
+          case "publish":
+            if (!this.isPublished) {
+              // show the plio preview
+              this.togglePlioPreviewMode();
+            }
+            break;
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Editor.vue:737]: created a task to do this in a separate PR

**Pull request merged**